File: src/api/project.ts
import { invoke } from '@tauri-apps/api/core'
import type { Project, Schematic } from '@/types'
import type { ProjectSettings } from '@/types/settings'

/**
 * Project API - Frontend interface to Rust backend project commands
 */
export const projectAPI = {
  /**
   * Create a new project
   */
  async createProject(name: string): Promise<Project> {
    return await invoke<Project>('create_project', { name })
  },

  /**
   * Open an existing project from file
   */
  async openProject(path: string): Promise<Project> {
    return await invoke<Project>('open_project', { path })
  },

  /**
   * Save the current project
   */
  async saveProject(project: Project, path: string): Promise<void> {
    return await invoke<void>('save_project', { project, path })
  },

  /**
   * Save project with a new path
   */
  async saveProjectAs(project: Project, newPath: string): Promise<void> {
    return await invoke<void>('save_project_as', { 
      project, 
      newPath 
    })
  },

  /**
   * Get list of recently opened projects
   */
  async getRecentProjects(): Promise<string[]> {
    return await invoke<string[]>('get_recent_projects')
  },

  /**
   * Add a new schematic to the project
   */
  async addSchematicToProject(
    projectId: string, 
    schematicName: string
  ): Promise<Schematic> {
    return await invoke<Schematic>('add_schematic_to_project', { 
      projectId, 
      schematicName 
    })
  },

  /**
   * Get project information by ID
   */
  async getProjectInfo(projectId: string): Promise<Project> {
    return await invoke<Project>('get_project_info', { projectId })
  },

  /**
   * Update project settings
   */
  async updateProjectSettings(
    projectId: string, 
    settings: ProjectSettings
  ): Promise<void> {
    return await invoke<void>('update_project_settings', { 
      projectId, 
      settings 
    })
  }
}

File: src-tauri/src/commands/project.rs
use crate::models::{Project, Schematic};
use crate::services::project_service::ProjectService;
use tauri::State;
use std::sync::Arc;

#[tauri::command]
pub async fn create_project(name: String) -> Result<Project, String> {
    if !crate::utils::validators::validate_project_name(&name) {
        return Err("Invalid project name".to_string());
    }
    
    Project::new(name)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn open_project(
    path: String,
    project_service: State<'_, Arc<ProjectService>>
) -> Result<Project, String> {
    if !crate::utils::validators::validate_file_path(&path) {
        return Err("Invalid file path".to_string());
    }
    
    project_service.load_project(&path)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn save_project(
    project: Project,
    path: String,
    project_service: State<'_, Arc<ProjectService>>
) -> Result<(), String> {
    if !crate::utils::validators::validate_file_path(&path) {
        return Err("Invalid file path".to_string());
    }
    
    project_service.save_project(&project, &path)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn save_project_as(
    project: Project,
    new_path: String,
    project_service: State<'_, Arc<ProjectService>>
) -> Result<(), String> {
    if !crate::utils::validators::validate_file_path(&new_path) {
        return Err("Invalid file path".to_string());
    }
    
    project_service.save_project(&project, &new_path)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_recent_projects(
    project_service: State<'_, Arc<ProjectService>>
) -> Result<Vec<String>, String> {
    Ok(project_service.get_recent_projects().await)
}

#[tauri::command]
pub async fn add_schematic_to_project(
    project_id: String,
    schematic_name: String,
    project_service: State<'_, Arc<ProjectService>>
) -> Result<Schematic, String> {
    if !crate::utils::validators::validate_string_length(&schematic_name, 100) {
        return Err("Invalid schematic name".to_string());
    }
    
    let schematic = Schematic::new(schematic_name)
        .map_err(|e| e.to_string())?;
    
    project_service.add_schematic(&project_id, schematic.clone())
        .await
        .map_err(|e| e.to_string())?;
    
    Ok(schematic)
}

#[tauri::command]
pub async fn get_project_info(
    project_id: String,
    project_service: State<'_, Arc<ProjectService>>
) -> Result<Project, String> {
    project_service.get_project(&project_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_project_settings(
    project_id: String,
    settings: serde_json::Value,
    project_service: State<'_, Arc<ProjectService>>
) -> Result<(), String> {
    project_service.update_settings(&project_id, settings)
        .await
        .map_err(|e| e.to_string())
}